File: velocity/src/main/java/fun/ksnb/multilogin/velocity/impl/VelocityPlayerManager.java
package fun.ksnb.multilogin.velocity.impl;

import com.velocitypowered.api.proxy.Player;
import com.velocitypowered.api.proxy.ProxyServer;
import moe.caa.multilogin.api.internal.plugin.IPlayer;
import moe.caa.multilogin.api.internal.plugin.IPlayerManager;

import java.util.Optional;
import java.util.Set;
import java.util.UUID;
import java.util.stream.Collectors;

/**
 * Velocity 玩家管理器对象
 */
public class VelocityPlayerManager implements IPlayerManager {
    private final ProxyServer server;

    public VelocityPlayerManager(ProxyServer server) {
        this.server = server;
    }

    @Override
    public Set<IPlayer> getPlayers(String name) {
        return server.getAllPlayers().stream().filter(p -> p.getUsername().equalsIgnoreCase(name)).map(VelocityPlayer::new).collect(Collectors.toSet());
    }

    @Override
    public IPlayer getPlayer(UUID uuid) {
        Optional<Player> player = server.getPlayer(uuid);
        return player.map(VelocityPlayer::new).orElse(null);
    }

    @Override
    public Set<IPlayer> getOnlinePlayers() {
        return server.getAllPlayers().stream().map(VelocityPlayer::new).collect(Collectors.toSet());
    }
}

File: velocity/velocity-core/src/main/kotlin/fun/iiii/multilogin/velocity/core/manager/VelocityPlayerManager.kt
package `fun`.iiii.multilogin.velocity.core.manager

import com.velocitypowered.api.proxy.Player
import `fun`.iiii.multilogin.velocity.core.main.MultiLoginVelocityCore
import `fun`.iiii.multilogin.velocity.core.util.toMultiLoginGameProfile
import moe.caa.multilogin.core.plugin.IPlayerManager
import net.kyori.adventure.audience.Audience
import net.kyori.adventure.text.Component
import java.util.*

class VelocityPlayerManager(private val core: MultiLoginVelocityCore) : IPlayerManager<Player> {
    override fun getPlayerType() = Player::class.java
    override fun getPlayer(audience: Audience) = VelocityPlayerInfo(audience as Player)
    override fun getOnlinePlayers() = core.bootstrap.proxyServer.allPlayers.map { VelocityPlayerInfo(it) }
    override fun getOnlinePlayer(uuid: UUID): VelocityPlayerInfo? =
        core.bootstrap.proxyServer.getPlayer(uuid).map { VelocityPlayerInfo(it) }.orElse(null)

    override fun getOnlinePlayer(string: String): IPlayerManager.IPlayerInfo? =
        core.bootstrap.proxyServer.getPlayer(string).map { VelocityPlayerInfo(it) }.orElse(null)

    override fun broadcastMessage(component: Component) =
        core.bootstrap.proxyServer.allPlayers.forEach { it.sendMessage(component) }


    class VelocityPlayerInfo(private val player: Player) : IPlayerManager.IPlayerInfo {
        override val audience = player
        override val inGameProfile = player.gameProfile.toMultiLoginGameProfile()

        override fun disconnect(component: Component) {
            player.disconnect(component)
        }

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (javaClass != other?.javaClass) return false
            other as VelocityPlayerInfo
            return player == other.player
        }

        override fun hashCode(): Int {
            return player.hashCode()
        }
    }
}